### PR TITLE
Align IP selection behavior with kernel

### LINKED ISSUE #22853: CFP: Improve `firstGlobalAddr` IP selection
## Cilium Feature Proposal

**Is your feature request related to a problem?**

Floating IP is being chosen as NodePort IP instead of primary IP on additional host interfaces.
Related to https://github.com/cilium/cilium/issues/15332, https://github.com/cilium/cilium/pull/15333. 

**(Optional) Describe your proposed solution**

Would like to implement @pchaigno's proposal in https://github.com/cilium/cilium/pull/15333#issuecomment-804840937:

- [ ] Fix `netlink.AddrList` [upstream](https://github.com/vishvananda/netlink/blob/c21bda41e995cbc0b91d34bec008a48f05f14d8b/addr_linux.go#L177) to sort by interface indexes (or create a new function).
- [ ] Remove the sort on addresses in our `firstGlobalAddr` function.
- [ ] Extend `firstGlobalAddr` match [the kernel's selection](https://elixir.bootlin.com/linux/v5.11.8/source/net/ipv4/devinet.c#L1318) and skip secondary addresses.

### PR DESCRIPTION
**Pull request by akhilles**
Sorting IPs by `bytes.Compare` gets rid of creation order, which can
lead to secondary IPs being chosed for masquerading and NodePort SNAT
purposes. Instead, stably sort by ifindex to preserve creation order and
give preference to primary IPs. Also, filter out secondary IPs.

This also more closely aligns with the Linux kernel's IP selection
behavior.

Fixes: #22853

```release-note
Align selection of IP addresses used for masquerading and NodePort SNAT with Linux kernel behavior, by preferring addresses assigned to the interface earlier and filtering out secondary addresses.
```

### CODE REVIEW

**Comment by sayboras** [comment]
Thanks for your PR. I am not familiar with this code path, so I will ask for  other agent member to review the changes :pray:.

**Comment by akhilles** [comment]
Would appreciate a review on this. Anything I can do to help?

File: pkg/node/address_linux.go [hunk 1]
```diff
@@ -42,6 +41,11 @@ func firstGlobalAddr(intf string, preferredIP net.IP, family int, preferPublic b
 
 retryInterface:
 	addr, err := netlink.AddrList(link, family)
+	// Ensure stable order so that the same IP is chosen every time.
+	// More info in the issue GH-7637.
+	sort.SliceStable(addr, func(i, j int) bool {
```
Comment (ldelossa) [inline pkg/node/address_linux.go]: What does this sort accomplish? `netlink.AddrList` is filtered to a `link` already, meaning all the `LinkIndex` fields on each address will be the same, producing the same list that `netlink.AddrList()` returned to begin with, no?

**Review by ldelossa — CHANGES_REQUESTED** [review]

File: pkg/node/address_linux.go (hunk 1 above)
Comment (ldelossa) [reply pkg/node/address_linux.go]: Scratch that, I see `netlink.AddrList` may get called with a nil link.

**Comment by ldelossa** [comment]
@akhilles I believe the comments on the caller to this function  `firstGlobalV4Addr` will need updating. Most notably, this statement is no longer true, correct?

```
// firstGlobalV4Addr returns the first IPv4 global IP of an interface,
// where the IPs are sorted in ascending order.
```

I believe the addresses will be returned in creation order, as that's how the kernel returns them on my machine.

File: pkg/node/address_linux.go [hunk 2]
```diff
@@ -57,16 +61,19 @@ retryScope:
...
-					ipsPublic = append(ipsPublic, a.IP)
-				} else {
-					ipsPrivate = append(ipsPrivate, a.IP)
-				}
 				// If the IP is the same as the preferredIP, that
 				// means that maybe it is restored from node_config.h,
 				// so if it is present we prefer this one.
 				if a.IP.Equal(preferredIP) {
 					hasPreferred = true
+				} else if a.Flags&unix.IFA_F_SECONDARY > 0 {
```
Comment (ldelossa) [inline pkg/node/address_linux.go]: If we want to filter our secondary addresses, I think we should do it all of the time. Hoisting this to the first or second statement in for loop maybe best.

Comment (akhilles) [reply pkg/node/address_linux.go]: In this case, I think `preferredIP` should take precedence over primary/secondary because it's being explicitly provided as an input.

I don't have a strong preference, but respecting `preferredIP` seems less surprising.

Comment (pchaigno) [reply pkg/node/address_linux.go]: Maybe we could refactor the code a bit to make this clearer (note the comment update as well):
```go
// If the IP is the same as the preferredIP, that
// means that maybe it is restored from node_config.h,
// so if it is present we prefer this one, even if it
// is a secondary address.
if a.IP.Equal(preferredIP) {
	hasPreferred = true
	continue
}

if a.Flags&unix.IFA_F_SECONDARY > 0 {
	// Skip secondary addresses
	continue
}
```

File: pkg/node/address_linux.go [hunk 3]
```diff
@@ -42,6 +41,11 @@ func firstGlobalAddr(intf string, preferredIP net.IP, family int, preferPublic b
 
 retryInterface:
 	addr, err := netlink.AddrList(link, family)
+	// Ensure stable order so that the same IP is chosen every time.
+	// More info in the issue GH-7637.
+	sort.SliceStable(addr, func(i, j int) bool {
+		return addr[i].LinkIndex < addr[j].LinkIndex
+	})
```
Comment (pchaigno) [inline pkg/node/address_linux.go]: Any reason to perform this sort before IPs are even filtered?

File: pkg/node/address_linux.go [hunk 4]
```diff
@@ -57,16 +61,19 @@ retryScope:
...
-					ipsPrivate = append(ipsPrivate, a.IP)
-				}
 				// If the IP is the same as the preferredIP, that
 				// means that maybe it is restored from node_config.h,
 				// so if it is present we prefer this one.
 				if a.IP.Equal(preferredIP) {
 					hasPreferred = true
+				} else if a.Flags&unix.IFA_F_SECONDARY > 0 {
+					// Skip secondary addresses
+					continue
```
Comment (pchaigno) [inline pkg/node/address_linux.go]: AFAICS, there are two independent changes here: the sort and the filter for secondary addresses. In your commit you mention that this fixes an issue with floating IPs in your case. Which change is fixing your issue exactly?

Given they are independent changes, I would also prefer to have them in separate commits. We may need to backport one change without backporting the other.

**Review by pchaigno — COMMENTED** [review]
> Previously, the IPs were sorted and the first value was chosen. This could cause a floating IP to be chosen for SNAT use instead of the primary IP address.
> 
> This patch stably sorts addresses by link index, which preserves the ordering of IPs for any given interface.

Could you give a bit more detail here? IIUC, it's the sort that is fixing your issue with floating IP addresses. How come? Is it because your floating IPs are always assigned to an interface that is created after the interface with the primary address?

> ```release-note
> Prefer primary IPs for masquerade, NodePort address selection
> ```

Maybe reword to:
```
Align selection of IP address for masquerading on Linux's logic, by preferring older interfaces and filtering out secondary addresses.
```

File: pkg/node/address_linux.go (hunk 2 above)
Comment (ldelossa) [reply pkg/node/address_linux.go]: Won't this be a bit problematic in the case where you may have a preferred IP on an interface, but also a secondary IP which overlaps it? You'd potentially select the secondary IP if it was created first I believe. 

Maybe these are edge cases we aren't concerned with?

File: pkg/node/address_linux.go (hunk 4 above)
Comment (akhilles) [reply pkg/node/address_linux.go]: The sort by creation time is enough to fix our issue, because the floating IP is assigned by an agent on the node after the primary IP has been assigned.

I'll separate them into separate commits.

File: pkg/node/address_linux.go (hunk 3 above)
Comment (akhilles) [reply pkg/node/address_linux.go]: I was thinking it helps with readability since we're only sorting in one place, right after fetching the addrs. I can switch back to sorting `ipsPublic` and `ipsPrivate` separately. Wdty about fetching addrs + filtering out secondary IPs + sorting in that order?

File: pkg/node/address_linux.go (hunk 2 above)
Comment (akhilles) [reply pkg/node/address_linux.go]: I might be misunderstanding the scenario you described, but I think the preferred IP would still be selected since it has precedence over the ordering *and* the filtering.

File: pkg/node/address_linux.go (hunk 3 above)
Comment (pchaigno) [reply pkg/node/address_linux.go]: Upside of sorting at the end is that we have less stuff to sort, though I'm not sure it makes for a significant difference.

File: pkg/node/address_linux.go (hunk 4 above)
Comment (pchaigno) [reply pkg/node/address_linux.go]: > The sort by creation time is enough to fix our issue, because the floating IP is assigned by an agent on the node after the primary IP has been assigned.

That's what I don't understand. If the floating IP is assigned to the same interface as the primary address, why does sorting by interface indexes change anything? 

> I'll separate them into separate commits.

Thanks!

File: pkg/node/address_linux.go (hunk 3 above)
Comment (akhilles) [reply pkg/node/address_linux.go]: I'll switch this back to sorting at the end since it's not a strictly related change.

File: pkg/node/address_linux.go (hunk 4 above)
Comment (akhilles) [reply pkg/node/address_linux.go]: Sorting by ifindex doesn't solve the problem, but removing the sort by `bytes.Compare` does. Sorting by `bytes.Compare` gets rid of creation order, which leads to secondary IPs being chosen instead of primary IPs. The sorting was added to address https://github.com/cilium/cilium/issues/7637#issuecomment-482152467, which actually mentions sorting by ifindex as a potential fix but the PR that resolved that issue ended up sorting using `bytes.Compare` instead. Maybe because ifindex wasn't available?

So, sorting by ifindex is an alternative to sorting by byte comparison that preserves creation order of IPs for any given interface.

I should probably clarify this in the commit message.

Comment (pchaigno) [reply pkg/node/address_linux.go]: Ah, I see that makes sense 👍 

Yep, an explanation of that in the commit description would be great!

File: pkg/node/address_linux.go (hunk 2 above)
Comment (ldelossa) [reply pkg/node/address_linux.go]: What I'm trying to get it is, we must explain this function like this, currently, "We filter secondary IP addresses, unless it is preferred. If two preferred IP addresses exist across interfaces, we return the one created first". 

If we always filter secondary, the outcome is more predictable. "We filter secondary IP addresses. IPs are returned in creation order across interfaces". 

Question tho, given your statement here: https://github.com/cilium/cilium/pull/22866#discussion_r1065007987 why do you propose we filter these secondary addresses? It seems like simply sorting by creation order fixes your issue?

Comment (akhilles) [reply pkg/node/address_linux.go]: Filtering secondary addresses isn't needed, I added it only to be consistent with the kernel's [`inet_select_addr`](https://elixir.bootlin.com/linux/v5.11.8/C/ident/inet_select_addr) function. Maybe it's better to remove it for now? I can't think of a use-case where creation order isn't sufficient for selecting a primary IP.

Comment (pchaigno) [reply pkg/node/address_linux.go]: I think being consistent with the kernel is good (as long as it's in a separate commit). That avoids surprises for users when they switch from kube-proxy to KPR for example.

Comment (akhilles) [reply pkg/node/address_linux.go]: I pulled the filtering into a separate commit and added this comment:
> All secondary IPs, except the preferredIP, are filtered out.

IMO, this is clear enough since currently, there's only one preferred IP that's provided as an input.

**Review by pchaigno — APPROVED** [review]
Thanks!

**Review by ldelossa — APPROVED** [review]
LGTM. Thanks.

**Comment by akhilles** [comment]
I think the test failure is unrelated or a flake. I noticed an unrelated PR (22768) failed with the similar error (on a different k8s version):
```
[unable to upgrade connection: Authorization error](error: unable to upgrade connection: Authorization error (user=kube-apiserver-kubelet-client, verb=create, resource=nodes, subresource=proxy))
```

Would it be possible to rerun the tests?

**Comment by pchaigno** [comment]
/test-1.25-4.19

**Comment by akhilles** [comment]
Looks like Cluster mesh w/ WireGuard test failure is also known flake: https://github.com/cilium/cilium/issues/23044.

**Comment by pchaigno** [comment]
> Looks like Cluster mesh w/ WireGuard test failure is also known flake: #23044.

Yep, that one was disabled in the meantime so fine to ignore.

Everything else is green and we have reviews so I'll mark ready to merge.

**Comment by akhilles** [comment]
Thanks!
